File: 2nd Year/2nd Semester/MPP/PROIECT-C#/Lab1/TriatlonNetworking/protobufprotocol/ProtoTriatlonWorker.cs
using System.Net.Sockets;
using Google.Protobuf;
using log4net;
using Triatlon.Network.ProtobufProtocol;
using TriatlonServicess;
using Arbitru = TriatlonModel.Arbitru;
using Participant = TriatlonModel.Participant;
using Rezultat = TriatlonModel.Rezultat;
using TipProba = TriatlonModel.TipProba;

namespace TriatlonNetworking.protobufprotocol
{
    public class ProtoTriatlonWorker : ITriatlonObserver
    {
        private readonly ITriatlonServices _server;
        private readonly TcpClient _connection;

        private NetworkStream _stream;
        private volatile bool _connected;

        private static readonly ILog Logger = LogManager.GetLogger(typeof(ProtoTriatlonWorker));

        public ProtoTriatlonWorker(ITriatlonServices server, TcpClient connection)
        {
            _server = server;
            _connection = connection;
            try
            {
                _stream = connection.GetStream();
                _connected = true;
            }
            catch (IOException e)
            {
                Logger.Error("Error creating worker: " + e.Message);
            }
        }

        public void Run()
        {
            while (_connected)
            {
                try
                {
                    // Read the request
                    Request request = Request.Parser.ParseDelimitedFrom(_stream);
                    if (request == null)
                    {
                        break;
                    }

                    Logger.Debug($"Received request: {request}");

                    // Process request and get response
                    Response response = HandleRequest(request);
                    if (response != null)
                    {
                        SendResponse(response);
                    }
                }
                catch (IOException e)
                {
                    Logger.Error($"Error processing request: {e.Message}");
                    _connected = false;
                }

                Thread.Sleep(10);
            }

            try
            {
                _stream.Close();
                _connection.Close();
            }
            catch (IOException e)
            {
                Logger.Error($"Error closing connection: {e.Message}");
            }
        }

        private void SendResponse(Response response)
        {
            Logger.Debug($"Sending response: {response}");
            lock (_stream)
            {
                response.WriteDelimitedTo(_stream);
                _stream.Flush();
            }
        }

        // ITriatlonObserver methods
        public void ArbitruLoggedIn(Arbitru arbitru)
        {
            Response response = ProtoUtils.CreateArbitruLoggedInResponse(arbitru);
            try
            {
                SendResponse(response);
            }
            catch (IOException e)
            {
                throw new TriatlonException("Error sending arbitru logged in notification: " + e.Message);
            }
        }

        public void ArbitruLoggedOut(Arbitru arbitru)
        {
            Response response = ProtoUtils.CreateArbitruLoggedOutResponse(arbitru);
            try
            {
                SendResponse(response);
            }
            catch (IOException e)
            {
                throw new TriatlonException("Error sending arbitru logged out notification: " + e.Message);
            }
        }

        public void RezultatAdded(Rezultat rezultat)
        {
            Response response = ProtoUtils.CreateAddedResultResponse(rezultat);
            try
            {
                SendResponse(response);
            }
            catch (IOException e)
            {
                throw new TriatlonException("Error sending rezultat added notification: " + e.Message);
            }
        }

        private static readonly Response OkResponse = ProtoUtils.CreateOkResponse();

        private Response HandleRequest(Request request)
        {
            switch (request.Type)
            {
                case Request.Types.Type.Login:
                    Logger.Debug($"Login request: {request.Arbitru}");
                    Arbitru loginArbitru = ProtoUtils.FromProto(request.Arbitru);
                    try
                    {
                        _server.Login(loginArbitru, this);
                        return OkResponse;
                    }
                    catch (TriatlonException e)
                    {
                        _connected = false;
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }

                case Request.Types.Type.Logout:
                    Logger.Debug($"Logout request: {request.Arbitru}");
                    Arbitru logoutArbitru = ProtoUtils.FromProto(request.Arbitru);
                    try
                    {
                        _server.Logout(logoutArbitru, this);
                        _connected = false;
                        return OkResponse;
                    }
                    catch (TriatlonException e)
                    {
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }

                case Request.Types.Type.Register:
                    Logger.Debug($"Register request: {request.Arbitru}");
                    Arbitru registerArbitru = ProtoUtils.FromProto(request.Arbitru);
                    try
                    {
                        bool success = _server.Register(
                            registerArbitru.Username,
                            registerArbitru.Password,
                            registerArbitru.FirstName,
                            registerArbitru.LastName
                        );
                        return success
                            ? OkResponse
                            : ProtoUtils.CreateErrorResponse("Registration failed - username already exists");
                    }
                    catch (TriatlonException e)
                    {
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }

                case Request.Types.Type.AddResult:
                    Logger.Debug("Add result request");
                    try
                    {
                        Participant participant = ProtoUtils.FromProto(request.Participant);
                        Arbitru arbitru = ProtoUtils.FromProto(request.Arbitru);

                        TipProba tipProba = _server.GetProbaForArbitru(arbitru);
                        int punctaj = request.Punctaj;

                        Logger.Debug(
                            $"participant: {participant}, arbitru: {arbitru}, tipProba: {tipProba}, punctaj: {punctaj}");

                        if (tipProba == null)
                        {
                            return ProtoUtils.CreateErrorResponse("TipProba is null");
                        }

                        _server.AddRezultat(participant, arbitru, tipProba, punctaj);
                        return OkResponse;
                    }
                    catch (TriatlonException e)
                    {
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }

                case Request.Types.Type.GetResultsForProba:
                    Logger.Debug("Get results for proba request");
                    try
                    {
                        TipProba tipProba = ProtoUtils.FromProto(request.TipProba);
                        var rezultate = _server.GetResultateForProba(tipProba);
                        return ProtoUtils.CreateGetResultsForProbaResponse(rezultate);
                    }
                    catch (TriatlonException e)
                    {
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }

                case Request.Types.Type.GetAllParticipants:
                    Logger.Debug("Get all participants request");
                    try
                    {
                        List<Participant> participants = _server.GetAllParticipants();
                        return ProtoUtils.CreateGetAllParticipantsResponse(participants);
                    }
                    catch (TriatlonException e)
                    {
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }

                case Request.Types.Type.CalculateTotalScore:
                    Logger.Debug("Calculate total score request");
                    try
                    {
                        Participant participant = ProtoUtils.FromProto(request.Participant);
                        int totalScore = _server.CalculateTotalScore(participant);
                        return ProtoUtils.CreateCalculateTotalScoreResponse(totalScore);
                    }
                    catch (TriatlonException e)
                    {
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }

                case Request.Types.Type.GetAllProbe:
                    Logger.Debug("Get all probe request");
                    try
                    {
                        var probe = _server.GetAllProbe();
                        return ProtoUtils.CreateGetAllProbeResponse(probe);
                    }
                    catch (TriatlonException e)
                    {
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }

                case Request.Types.Type.GetProbaForArbitru:
                    Logger.Debug("Get proba for arbitru request");
                    try
                    {
                        Arbitru arbitru = ProtoUtils.FromProto(request.Arbitru);
                        TipProba tipProba = _server.GetProbaForArbitru(arbitru);
                        return ProtoUtils.CreateGetProbaForArbitruResponse(tipProba);
                    }
                    catch (TriatlonException e)
                    {
                        return ProtoUtils.CreateErrorResponse(e.Message);
                    }

                case Request.Types.Type.FindArbitru:
                    Logger.Debug("Find arbitru request");
                    var arbitruDTO = request.Arbitru;
                    Logger.Debug($"Username: {arbitruDTO.Username}, Password: {arbitruDTO.Password}");

                    Arbitru foundArbitru = _server.FindArbitruByUsernameAndPassword(
                        arbitruDTO.Username,
                        arbitruDTO.Password);

                    if (foundArbitru != null)
                    {
                        return ProtoUtils.CreateFindArbitruResponse(foundArbitru);
                    }
                    else
                    {
                        return ProtoUtils.CreateErrorResponse("Invalid credentials");
                    }


                case Request.Types.Type.GetAllResults:
                    Logger.Debug("Get all results request");
                    var allResults = _server.GetAllResults();
                    return ProtoUtils.CreateGetAllResultsResponse(allResults);

                default:
                    return ProtoUtils.CreateErrorResponse("Invalid request type");
            }
        }
    }
}

File: 2nd Year/2nd Semester/MPP/PROIECT-JAVA/Laborator/TriatlonNetworking/src/main/java/triatlon/network/protobuffprotocol/ProtoTriatlonWorker.java
package triatlon.network.protobuffprotocol;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import triatlon.model.*;
import triatlon.services.ITriatlonObserver;
import triatlon.services.ITriatlonServices;
import triatlon.services.TriatlonException;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.List;

public class ProtoTriatlonWorker implements Runnable, ITriatlonObserver {
    private final ITriatlonServices server;
    private final Socket connection;

    private InputStream input;
    private OutputStream output;
    private volatile boolean connected;

    private static final Logger logger = LogManager.getLogger(ProtoTriatlonWorker.class);

    public ProtoTriatlonWorker(ITriatlonServices server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try {
            output = connection.getOutputStream();
            input = connection.getInputStream();
            connected = true;
        } catch (IOException e) {
            logger.error("Error creating worker: " + e.getMessage());
        }
    }

    @Override
    public void run() {
        while(connected) {
            try {
                // Read the request
                TriatlonProtobufs.Request request = TriatlonProtobufs.Request.parseDelimitedFrom(input);
                if (request == null) {
                    break;
                }

                logger.debug("Received request: {}", request);

                // Process request and get response
                TriatlonProtobufs.Response response = handleRequest(request);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (IOException e) {
                logger.error("Error processing request: {}", e.getMessage());
                connected = false;
            }
        }

        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            logger.error("Error closing connection: {}", e.getMessage());
        }
    }

    private void sendResponse(TriatlonProtobufs.Response response) throws IOException {
        logger.debug("Sending response: {}", response);
        synchronized (output) {
            response.writeDelimitedTo(output);
            output.flush();
        }
    }

    // ITriatlonObserver methods
    @Override
    public void arbitruLoggedIn(Arbitru arbitru) throws TriatlonException {
        TriatlonProtobufs.Response response = ProtoUtils.createArbitruLoggedInResponse(arbitru);
        try {
            sendResponse(response);
        } catch (IOException e) {
            throw new TriatlonException("Error sending arbitru logged in notification: " + e.getMessage());
        }
    }

    @Override
    public void arbitruLoggedOut(Arbitru arbitru) throws TriatlonException {
        TriatlonProtobufs.Response response = ProtoUtils.createArbitruLoggedOutResponse(arbitru);
        try {
            sendResponse(response);
        } catch (IOException e) {
            throw new TriatlonException("Error sending arbitru logged out notification: " + e.getMessage());
        }
    }

    @Override
    public void rezultatAdded(Rezultat rezultat) throws TriatlonException {
        TriatlonProtobufs.Response response = ProtoUtils.createRezultatAddedResponse(rezultat);
        try {
            sendResponse(response);
        } catch (IOException e) {
            throw new TriatlonException("Error sending rezultat added notification: " + e.getMessage());
        }
    }

    private static final TriatlonProtobufs.Response okResponse = ProtoUtils.createOkResponse();

    private TriatlonProtobufs.Response handleRequest(TriatlonProtobufs.Request request) {
        switch (request.getType()) {
            case LOGIN:
                logger.debug("Login request: {}", request.getArbitru());
                Arbitru loginArbitru = ProtoUtils.getArbitru(request.getArbitru());
                try {
                    server.login(loginArbitru, this);
                    return okResponse;
                } catch (TriatlonException e) {
                    connected = false;
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }

            case LOGOUT:
                logger.debug("Logout request: {}", request.getArbitru());
                Arbitru logoutArbitru = ProtoUtils.getArbitru(request.getArbitru());
                try {
                    server.logout(logoutArbitru, this);
                    connected = false;
                    return okResponse;
                } catch (TriatlonException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }

            case REGISTER:
                logger.debug("Register request: {}", request.getArbitru());
                Arbitru registerArbitru = ProtoUtils.getArbitru(request.getArbitru());
                try {
                    boolean success = server.register(
                            registerArbitru.getUsername(),
                            registerArbitru.getPassword(),
                            registerArbitru.getFirst_name(),
                            registerArbitru.getLast_name()
                    );
                    if (success) {
                        return okResponse;
                    } else {
                        return ProtoUtils.createErrorResponse("Registration failed - username already exists");
                    }
                } catch (TriatlonException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }

            case ADD_RESULT:
                logger.debug("Add result request");
                try {
                    Participant participant = ProtoUtils.getParticipant(request.getParticipant());
                    Arbitru arbitru = ProtoUtils.getArbitru(request.getArbitru());

                    TipProba tipProba = server.getProbaForArbitru(arbitru);
                    int punctaj = request.getPunctaj();

                    logger.debug("participant: {}, arbitru: {}, tipProba: {}, punctaj: {}", participant, arbitru, tipProba, punctaj);

                    if (tipProba == null) {
                        return ProtoUtils.createErrorResponse("TipProba is null");
                    }

                    server.addRezultat(participant, arbitru, tipProba, punctaj);
                    return okResponse;
                } catch (TriatlonException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }

            case GET_RESULTS_FOR_PROBA:
                logger.debug("Get results for proba request");
                try {
                    TipProba tipProba = ProtoUtils.getTipProba(request.getTipProba());
                    List<Rezultat> rezultate = server.getResultateForProba(tipProba);
                    return ProtoUtils.createGetResultsForProbaResponse(rezultate);
                } catch (TriatlonException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }

            case GET_ALL_PARTICIPANTS:
                logger.debug("Get all participants request");
                try {
                    List<Participant> participants = server.getAllParticipants();
                    return ProtoUtils.createGetAllParticipantsResponse(participants);
                } catch (TriatlonException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }

            case CALCULATE_TOTAL_SCORE:
                logger.debug("Calculate total score request");
                try {
                    Participant participant = ProtoUtils.getParticipant(request.getParticipant());
                    int totalScore = server.calculateTotalScore(participant);
                    return ProtoUtils.createCalculateTotalScoreResponse(totalScore);
                } catch (TriatlonException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }

            case GET_ALL_PROBE:
                logger.debug("Get all probe request");
                try {
                    List<Proba> probe = server.getAllProbe();
                    return ProtoUtils.createGetAllProbeResponse(probe);
                } catch (TriatlonException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }

            case GET_PROBA_FOR_ARBITRU:
                logger.debug("Get proba for arbitru request");
                try {
                    Arbitru arbitru = ProtoUtils.getArbitru(request.getArbitru());
                    TipProba tipProba = server.getProbaForArbitru(arbitru);
                    return ProtoUtils.createGetProbaForArbitruResponse(tipProba);
                } catch (TriatlonException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }

            case FIND_ARBITRU:
                logger.debug("Find arbitru request");
                TriatlonProtobufs.Arbitru arbitruDTO = request.getArbitru();
                logger.debug("Username: {}, Password: {}", arbitruDTO.getUsername(), arbitruDTO.getPassword());

                Arbitru foundArbitru = server.findArbitruByUsernameAndPassword(
                        arbitruDTO.getUsername(),
                        arbitruDTO.getPassword());

                if (foundArbitru != null) {
                    return ProtoUtils.createFindArbitruResponse(foundArbitru);
                } else {
                    return ProtoUtils.createErrorResponse("Invalid credentials");
                }

            case GET_ALL_RESULTS:
                logger.debug("Get all results request");
                List<Rezultat> allResults = server.getAllResults();
                return ProtoUtils.createGetAllResultsResponse(allResults);

            default:
                return ProtoUtils.createErrorResponse("Invalid request type");
        }
    }
}